driver-core/src/main/com/mongodb/internal/connection/StreamFactoryHelper.java

```diff
@@ -17,26 +17,67 @@
+import com.mongodb.connection.AsyncTransportSettings;
 import com.mongodb.connection.NettyTransportSettings;
+import com.mongodb.connection.SocketSettings;
 import com.mongodb.connection.TransportSettings;
+import com.mongodb.internal.ValueOrExceptionContainer;
 import com.mongodb.internal.connection.netty.NettyStreamFactoryFactory;
 import com.mongodb.spi.dns.InetAddressResolver;
 
+import java.nio.channels.AsynchronousChannelGroup;
+import java.util.concurrent.ExecutorService;
+
 /**
  * <p>This class is not part of the public API and may be removed or changed at any time</p>
  */
 public final class StreamFactoryHelper {
-    public static StreamFactoryFactory getStreamFactoryFactoryFromSettings(final TransportSettings transportSettings,
+
+    public static StreamFactory getSyncStreamFactory(final MongoClientSettings settings,
+            final InetAddressResolver inetAddressResolver, final SocketSettings socketSettings) {
+        TransportSettings transportSettings = settings.getTransportSettings();
+        if (transportSettings == null) {
+            return new SocketStreamFactory(inetAddressResolver, socketSettings, settings.getSslSettings());
+        } else if (transportSettings instanceof AsyncTransportSettings) {
+            throw new MongoClientException("Unsupported transport settings in sync: " + transportSettings.getClass().getName());
+        } else if (transportSettings instanceof NettyTransportSettings) {
+            return getNettyStreamFactoryFactory(inetAddressResolver, (NettyTransportSettings) transportSettings)
+                    .create(socketSettings, settings.getSslSettings());
+        } else {
+            throw new MongoClientException("Unsupported transport settings: " + transportSettings.getClass().getName());
+        }
+    }
+
+    public static StreamFactoryFactory getAsyncStreamFactoryFactory(final MongoClientSettings settings,
             final InetAddressResolver inetAddressResolver) {
-        if (transportSettings instanceof NettyTransportSettings) {
-            return NettyStreamFactoryFactory.builder().applySettings((NettyTransportSettings) transportSettings)
-                    .inetAddressResolver(inetAddressResolver)
-                    .build();
+        TransportSettings transportSettings = settings.getTransportSettings();
+        if (transportSettings == null || transportSettings instanceof AsyncTransportSettings) {
+            ExecutorService executorService = transportSettings == null
+                    ? null
+                    : ((AsyncTransportSettings) transportSettings).getExecutorService();
+            if (settings.getSslSettings().isEnabled()) {
+                return new TlsChannelStreamFactoryFactory(inetAddressResolver, executorService);
+            } else {
+                ValueOrExceptionContainer<AsynchronousChannelGroup> group = new ValueOrExceptionContainer<>(
+                        () -> AsynchronousChannelGroup.withThreadPool(executorService));
```

Comment:
The exception here is deferred until:
```
attemptConnectionChannel = AsynchronousSocketChannel.open(group.get());
```
This seemed like the cleanest alternative.

Note that in the [original code](https://github.com/mongodb/mongo-java-driver/commit/19336af05937e8a986cc583af6f86688bedf32a6#diff-728c81f23fa3ca6d19e4df9a5eaec1e997645e323e83d1c216986a603e878bceR48), the group itself was provided to the driver.

Reply:
Why this deferring was implemented? It's especially surprising that `AsynchronousChannelGroup.withThreadPool` is called here synchronously, but its exception is then kept "deferred" until `MongoClient` starts establishing connections to a MongoDB deployment.

Reply:
I removed it, without obvious issues. I do not recall why I considered this cleaner.